### PR TITLE
Async search: create internal index only before storing initial response

### LINKED ISSUE #54180: [CI] AsyncSearchActionTests.testCancellation fails on CI
The test `AsyncSearchActionTests.testCancellation` failed on CI on the 7.x branch:

```
org.elasticsearch.xpack.search.AsyncSearchActionTests > testCancellation FAILED
    java.lang.AssertionError: Shard [.async-search][0] is still locked after 5 sec waiting
        at __randomizedtesting.SeedInfo.seed([DD4D761E3C3EE063:B7A2968A281D7E5]:0)
        at org.junit.Assert.fail(Assert.java:88)
        at org.elasticsearch.test.InternalTestCluster.assertAfterTest(InternalTestCluster.java:2483)
        at org.elasticsearch.test.ESIntegTestCase.afterInternal(ESIntegTestCase.java:600)
        at org.elasticsearch.test.ESIntegTestCase.cleanUpCluster(ESIntegTestCase.java:2143)
```
Build scan: https://gradle-enterprise.elastic.co/s/4ksqilghmk6ii

This is the first time this test fails on CI. I'm not sure this is really related to async search as the logs mentioned some version conflict exception:

```
[2020-03-25T03:22:06,439][ERROR][o.e.x.s.TransportSubmitAsyncSearchAction] [node_s0] failed to store async-search [FkdmWEtzRnhjUVZ1cE9ya090RUE1TVEaWWR1SU5FRV9RcTZuYTBqa0kyYVVDZzo1NTQ=]
  org.elasticsearch.transport.RemoteTransportException: [node_s0][127.0.0.1:63916][indices:data/write/update[s]]
  Caused by: org.elasticsearch.index.engine.VersionConflictEngineException: [GfXKsFxcQVupOrkOtEA5MQ]: version conflict, required seqNo [0], primary term [1]. but no document was found
       at org.elasticsearch.index.engine.InternalEngine.planIndexingAsPrimary(InternalEngine.java:1065) ~[elasticsearch-7.7.0-SNAPSHOT.jar:7.7.0-SNAPSHOT]
```

It does not reproduce locally with:

```
./gradlew ':x-pack:plugin:async-search:test' --tests "org.elasticsearch.xpack.search.AsyncSearchActionTests.testCancellation" \
  -Dtests.seed=DD4D761E3C3EE063 \
  -Dtests.security.manager=true \
  -Dtests.locale=es-NI \
  -Dtests.timezone=Atlantic/Faroe \
  -Dcompiler.java=13
```

This also looks different from #53360.

### PR DESCRIPTION
**Pull request by javanna**
We currently create the .async-search index if necessary before performing any action (index, update or delete). Truth is that this is needed only before storing the initial response. The other operations are either update or delete, which will anyways not find the document to update/delete even if the index gets created when missing. This also caused `testCancellation` failures as we were trying to delete the document twice from the .async-search index, once from `TransportDeleteAsyncSearchAction` and once as a consequence of the search task being completed. The latter may be called after the test is completed, but before the cluster is shut down and causing problems to the after test checks, for instance if it happens after all the indices have been cleaned up. It is totally fine to try to delete a response that is no longer found, but not quite so if such call will also trigger an index creation.

With this commit we remove all the calls to createIndexIfNecessary from the update/delete operation, and we leave one call only from storeInitialResponse which is where the index is expected to be created.

Closes #54180

### CODE REVIEW

File: x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchIndexService.java [hunk 1]
```diff
@@ -194,31 +192,16 @@ void updateExpirationTime(String docId,
...
-                            listener.onFailure(new ResourceNotFoundException(searchId.getEncoded()));
-                        } else {
-                            listener.onResponse(new AcknowledgedResponse(true));
-                        }
-                    },
-                    exc -> {
-                        logger.error(() -> new ParameterizedMessage("failed to clean async-search [{}]", searchId.getEncoded()), exc);
-                        listener.onFailure(exc);
-                    })),
-                listener::onFailure));
```
Comment (javanna) [inline x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchIndexService.java:221]: I tried to make this logic more readable. I found the boolean flag hard to reason about especially as it was provided true only once. I moved the listener wrapping to the callers, where each caller needs to do something different.

File: x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportGetAsyncSearchAction.java [hunk 2]
```diff
@@ -57,8 +59,12 @@ protected void doExecute(Task task, GetAsyncSearchAction.Request request, Action
                         ActionListener.wrap(
                             p -> getSearchResponseFromTask(searchId, request, nowInMillis, expirationTime, listener),
                             exc -> {
-                                if (exc.getCause() instanceof DocumentMissingException == false) {
-                                    logger.error("failed to retrieve " + searchId.getEncoded(), exc);
+                                //don't even log when: the async search document or its index is not found. That can happen if an invalid
+                                //search id is provided and no async search initial response has been stored yet.
+                                if (exc.getCause() instanceof DocumentMissingException == false
+                                    && exc.getCause() instanceof IndexNotFoundException == false) {
```
Comment (javanna) [inline x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportGetAsyncSearchAction.java:65]: I was wondering if we are sure about exc.getCause here. What's the top level exception?

File: x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchIndexService.java (hunk 1 above)
Comment (jimczi) [reply x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchIndexService.java:221]: ++

File: x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportGetAsyncSearchAction.java (hunk 2 above)
Comment (jimczi) [reply x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportGetAsyncSearchAction.java:65]: A `RemoteTransportException`.  Maybe we can replace the instanceof with `ExceptionsHelper.status(exc) != RestStatus.NOT_FOUND` ?

File: x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportDeleteAsyncSearchAction.java [hunk 3]
```diff
@@ -58,15 +67,40 @@ protected void doExecute(Task task, DeleteAsyncSearchAction.Request request, Act
...
         AsyncSearchTask task = store.getTask(taskManager, searchId);
         if (task != null) {
-            task.cancelTask(() -> store.deleteResponse(searchId, false, listener));
+            //the task was found and gets cancelled. The response may or may not be found, but we will return 200 anyways.
+            task.cancelTask(() -> store.deleteResponse(searchId,
+                ActionListener.wrap(
+                    r -> listener.onResponse(new AcknowledgedResponse(true)),
+                    exc -> {
+                        //the index may not be there (no initial async search response stored yet?): we still want to return 200
+                        if (exc.getCause() instanceof IndexNotFoundException) {
```
Comment (jimczi) [inline x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportDeleteAsyncSearchAction.java:79]: We shouldn't fail If the document is missing:
```suggestion
                        if (ExceptionsHelper.status(exc) == RestStatus.NOT_FOUND) {
```

**Review by jimczi — COMMENTED** [review]
This makes sense to me. I left some comments.

File: x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportDeleteAsyncSearchAction.java (hunk 3 above)
Comment (javanna) [reply x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportDeleteAsyncSearchAction.java:79]: we don't, document missing does not come back as a failure. Yet, I agree on changing the condition as exc.getCause() instanceof IndexNotFoundException is not great

File: x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportGetAsyncSearchAction.java (hunk 2 above)
Comment (javanna) [reply x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportGetAsyncSearchAction.java:65]: the problem I have with `ExceptionsHelper.status(exc)` is that it does not look at the cause at all and it gives 500 if it does not know any better. How about `ExceptionsHelper.status(ExceptionsHelper.unwrapCause(exc))` ?

Comment (jimczi) [reply x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportGetAsyncSearchAction.java]: +1

File: x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportDeleteAsyncSearchAction.java (hunk 3 above)
Comment (jimczi) [reply x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/TransportDeleteAsyncSearchAction.java]: ++, thanks

**Pull request merged**